docker/Dockerfile

```diff
@@ -0,0 +1,159 @@
+#    changes. If we do `COPY . /app` now, this will rebuild when *any* file changes.
+#
+# 2) Ensures we *have* to mount the volume for dev image, as there's no embedded
+#    version of the code. Otherwise, we could end up accidentally using the
+#    version of the code included when the prod image was built.
+FROM base-python as job-runner-base
+
+# Create a non-root user to run the app as
+# We currently use host bind mounts in production backends, which means if we
+# need to match the uid inside the container to the one on the host.  To do
+# this, we reserve uid/gid 10000 to use as our running user on the host.
+RUN useradd --create-home --user-group --uid 10000 appuser
+
+# copy venv over from builder image. These will have root:root ownership, but
+# are readable by all.
+COPY --from=builder /opt/venv /opt/venv
+
+# Ensure we're using the venv by default
+ENV VIRTUAL_ENV=/opt/venv/ PATH="/opt/venv/bin:$PATH"
+
+RUN mkdir /app
+WORKDIR /app
+VOLUME /app
+
+# We set command rather than entrypoint, to make it easier to run different
+# things from the cli
+CMD ["/opt/venv/bin/python", "-m", "jobrunner.service"]
+
+# This may not be necessary, but it probably doesn't hurt
+ENV PYTHONPATH=/app
+
+# switch to running as the user
+USER appuser
+
+
+##################################################
+#
+# Production image
+#
+# Copy code in, add proper metadata
+FROM job-runner-base as job-runner-prod
+
+# Adjust this metadata to fit project. Note that the base-docker image does set
+# some basic metadata.
+LABEL org.opencontainers.image.title="job-runner" \
+      org.opencontainers.image.description="OpenSAFELY secure job-runner service" \
+      org.opencontainers.image.source="https://github.com/opensafely-core/job-runner"
+
+# copy application code
+COPY . /app
+
+# finally, tag with build information. These will change regularly, therefore
+# we do them as the last action.
+ARG BUILD_DATE=unknown
+LABEL org.opencontainers.image.created=$BUILD_DATE
+ARG GITREF=unknown
+LABEL org.opencontainers.image.revision=$GITREF
+
+
+
+##################################################
+#
+# Dev image
+#
+# Now we build a dev image from our job-runner-dev image. This is basically
+# installing dev dependencies and matching local UID/GID. It is expected that
+# the current code will be mounted in /app when this is run
+#
+FROM job-runner-base as job-runner-dev
+
+# switch back to root to run the install of dev requirements.txt
+USER root
+
+# TODO: its possible python dev dependencies might need some additional build packages installed?
+
+# install development requirements
+COPY requirements.dev.txt /tmp/requirements.dev.txt
+# using cache mount instead
+# hadolint ignore=DL3042
+RUN --mount=type=cache,target=/root/.cache \
+    python -m pip install --requirement /tmp/requirements.dev.txt
+
+# in dev, ensure appuser uid matches host user id
+ARG USERID=1000
+ARG GROUPID=1000
+RUN usermod -u $USERID appuser && groupmod -g $GROUPID appuser
```

Comment:
so - this fails for me on macOS, but release-hatch also fails in the same way - not sure if you want to fix now or update them both later. Failure is 
```
[job-runner-dev 3/3] RUN usermod -u 501 appuser && groupmod -g 20 appuser:
#25 0.231 groupmod: GID '20' already exists
```
which are my user (`501`) and first group in my groups listing, which is `staff` aka `20`. not a blocker for this PR, I think.

Reply:
Well, that is a pain. I'll add an issue to fix, somehow